pypeit/scripts/collate_1d.py

```diff
@@ -0,0 +1,817 @@
+        spec2d_files (:obj:`list` of str): 
+            List of spec2d files to build the map from.
+        par (:obj:`Collate1DPar):
+            Parameters from .collate1d file
+
+    Return:
+        :obj:`dict` Mapping of slit mask ids to the flags that caused the
+        slit to be excluded.
+    """
+
+    # Get the types of slits to exclude from our parameters
+    exclude_flags = par['collate1d']['slit_exclude_flags']
+    if isinstance(exclude_flags, str):
+        exclude_flags = [exclude_flags]
+
+    # Go through the slit_info of all spec2d files and find
+    # which slits should be excluded based on their flags
+    bit_mask = SlitTraceBitMask()
+    exclude_map = dict()
+    for spec2d_file in spec2d_files:
+
+        allspec2d = AllSpec2DObj.from_fits(spec2d_file)
+        for sobj2d in [allspec2d[det] for det in allspec2d.detectors]:
+            for (slit_id, mask, slit_mask_id) in sobj2d['slits'].slit_info:
+                for flag in exclude_flags:
+                    if bit_mask.flagged(mask, flag):
+                        if slit_mask_id not in exclude_map:
+                            exclude_map[slit_mask_id] = {flag}
+                        else:
+                            exclude_map[slit_mask_id].add(flag)
+
+    return exclude_map
+
+def group_spectra_by_source(spec1d_files, exclude_map, match_type, tolerance, unit=u.arcsec):
+    """Given a list of spec1d files from PypeIt, group the spectra within the
+    files by their source object. The grouping is done by comparing the 
+    position of each spectra (using either pixel or RA/DEC) using a given tolerance.
+
+    Args:
+        spec1d_files (list of str): A list of spec1d files created by PypeIt.
+        exclude_map (dict): Mapping of excluded slit mask ids to the reason
+            they were excluded.  Any spectra for these slits will be ignored.
+        match_type (str): How to match the positions of spectra. 'pixel' for
+            matching by spatial pixel distance or 'ra/dec' for matching by
+            angular distance.
+        tolerance (float): 
+            Maximum distance that two spectra can be from each other to be 
+            considered to be from the same source. Measured in floating
+            point pixels or as an angular distance (see ``unit1`` argument).
+        unit (:obj:`astropy.units.Unit`):
+            Units of ``tolerance`` argument if match_type is 'ra/dec'. 
+            Defaults to arcseconds. Igored if match_type is 'pixel'.
+
+    Returns (list of `obj`:SourceObject): The grouped spectra as SourceObjects.
+
+    """
+    source_list = []
+
+    for spec1d_file in spec1d_files:
+        sobjs = specobjs.SpecObjs.from_fitsfile(spec1d_file)
+
+        for sobj in sobjs.specobjs:
+
+            if sobj.MASKDEF_ID in exclude_map:
+                msgs.info(f'Excluding {sobj.MASKDEF_ID} in {spec1d_file} because of flags {exclude_map[sobj.MASKDEF_ID]}')
+                continue
+            if sobj.OPT_COUNTS is None:
```

Comment:
I added BOX_COUNTS to this check, so that only objects with no counts are excluded. With this change BOXCAR only can be done by changing ex_value in the coadd1d parameters. But it's still not possible to mix BOX and OPT in the same run of collate_pypeit_1d. Would we ever want to do that?

Reply:
good question.  I'll say no for now. :)